### ISSUE TITLE
Query about the reproducibility of Motion Capture datset in "Scalable Gradients..." (Li et al., 2020)

### ISSUE THREAD

**Issue by nghiahhnguyen** [google-research/torchsde#112]
I am trying to reproduce the results of the CMU Motion Capture dataset. I use references from examples/latent_sde_lorenz.py, the paper, and the preprocessed dataset [linked by ODE2VAE's repo](https://github.com/cagatayyildiz/ODE2VAE).

My current runs' results has large discrepancy from the results in the paper so I want to check if there are any training details I'm missing. (I am not very familiar with Bayesian modelling so I try to follow the hyperparameters in the paper as closely as possible.)

Here are the main issues:
- The validation and the test MSE for Latent SDE is in the range of 20-30, while Latent SDE's Test MSE in the paper is 4.03 +- 0.20.
- The log likelihood term in the ELBO for Latent SDE is in the magnitude of ```10^6``` to ```10^9``` depending on the choice of hyperparameters, while the code for ODE2VAE has the log likelihood for ODE2VAE in the magnitude of ```10^4```.

Here are the main training details I think I am most likely to wrongly interpret from the paper:
- The **standard deviation of the observation noise** when calculating the log likelihood (as in [this example](https://github.com/google-research/torchsde/blob/master/examples/latent_sde_lorenz.py#L192). I'm choosing this std to be ```0.01``` like in the example. I also tried varying this to ```0.1``` or ```0.001``` to no success. This parameter is not mentioned in the paper so I'm not sure what a sensible value is.
- When **calculating the log likelihood**, I am following [your example](https://github.com/google-research/torchsde/blob/master/examples/latent_sde_lorenz.py#L193) and take sum over the all timesteps and the data dimension, then take the mean over the number of samples. Please let me know if this correct for the CMU mocap dataset.
- In **training**, the prediction and the **log likelihood** should only be calculated for the last 297 predictions (since the first 3 observations are used to encode the context).
- The **solver** used is not mentioned. I have tried ```euler_heun```, ```milstein```, and even ```reversible_heun```.
- The **initial learning rate** for Adam is ```0.01``` in the paper, but in some runs with the initial lr to be ```0.001```, they seem to be more stable. I'm curious if you have any comments about this.
- The **dt for the solver** is said to be ```1/5``` of the minimum time difference between two observations. All observations are regular so we can choose the minimum time to be a particular value ```a``` (eg., 1), then ```dt``` would be ```0.2 * a```. I want to know if my interpretation here is correct. The paper didn't mention this value ```a``` or the start/end time. It would be nice if you remembered this.

Tagging @lxuechen since you know the most about the exp details. Thank you for your time!

**Comment by nghiahhnguyen** [comment]
After looking further, I find that my reproduction lacks the KL penalty term (mentioned as the BNN weight penalty in the ODE2VAE paper). The details on how exactly you calculate this term are not mentioned in the paper (e.g whether to calculate the KL penalty at ```t0``` or ```tN```). I would be grateful if you can elaborate further on this.

**Comment by lxuechen** [comment]
> The initial learning rate for Adam is 0.01 in the paper, but in some runs with the initial lr to be 0.001, they seem to be more stable. I'm curious if you have any comments about this. 

0.01 worked for my experiments if you apply reasonable decay. 

> After looking further, I see that the KL penalty at t0 (KL divergence between p(z) and q(z)) is already included. I am wondering if we also need the KL penalty at tN (KL divergence between p(x|z) and q(x|z))?

You would need KL at time 0, and the KL between the two stochastic processes. 

A nice example of how this is computed can be found [here](https://github.com/google-research/torchsde/blob/53038a3efcd77f6c9f3cfd0310700a59be5d5d2d/examples/latent_sde_lorenz.py#L198) for a toy task. `logqp_path` is what you're looking for. `logqp0` is the KL at time 0.

**Comment by abdulfatir** [comment]
@nghiahhnguyen were you able to reproduce the results for this dataset?
@lxuechen Could you please clarify which KL reweighting coefficient from {1, 0.1, 0.01, 0.001} worked best in your experiments (for LatentODE and LatentSDE)? Also, was this with or without the linear annealing schedule?

**Comment by nghiahhnguyen** [comment]
I was able to reduce the Test MSE to the range of 8.x - 1x.x after some changes, which is much lower than my first attempt, but still short of the figure 4.03 +- 0.20 in the paper.

**Comment by nghiahhnguyen** [comment]
Hi @abdulfatir, one point I found the most helpful, which I overlooked in my first attempt, is that the standard deviation of the predicted observations should be learnable (as stated in the paper), but the [example](https://github.com/google-research/torchsde/blob/53038a3efcd77f6c9f3cfd0310700a59be5d5d2d/examples/latent_sde_lorenz.py#L192) used a fixed hyperparameter. When I noticed and changed it, Test MSE dropped significantly to [8.x, 1x.x], as I mentioned above. I'm not sure if you have noticed it already but I guess it's worth mentioning!

Other than that, I don't recall any significant changes I made. In case it helps, I'm listing some significant hyperparameters in my best run:
- Use the adjoint method.
- Solvers: reversible_heun & adjoint_reversible_heun.
- Step size: solve from t0=0.3 to t1=30 with a dt_ratio of 0.2 (I have no specific reasoning for the choice of t0 and t1, just a random thing I tried).
- Beta (the coefficient of the KL divergence in the training loss): 1 with a linear schedule of 400 epochs (Looking back at the Val plot, the Val MSE loss starts dropping significantly lower than previous runs at around epoch 500. Val MSE loss starts plateauing around epoch 670. Maybe this is an interesting correlation!)
- Gradient clipping: 0.5 (I'm using PyTorch Lightning, so this is done with their built-in implementation).

With the above setting, my log-likelihood is around x*1e4 for training and x*1e4 for validation and testing. Please let me know if you managed to make further progress!

**Comment by abdulfatir** [comment]
Thanks a lot @nghiahhnguyen! I did use trainable scales in my experiments but only as a trainable vector (not output from a NN as in the paper).

Thanks for sharing your hyperparameters. I am giving it a go now. Will keep you posted. Are you sure about clipping the grad norm to 0.5? Given the magnitude of gradients for this model, this is quite a small value. 

Something I noticed when using scales output from a NN for the observation distribution is that the model becomes very susceptible to crashing with NaNs (possibly due to numerical errors). Did you experience something similar?

**Comment by nghiahhnguyen** [comment]
I'm glad that I can be of help @abdulfatir. 

I also experienced a lot of crashing with NaNs. That's why I started with such a small value of max norm for gradient clipping. I'm using gradient clipping only to stabilize the training process, so I guess you can try larger values and see if the process is still stable. I remembered that trying much larger values does not prevent NaNs-related crashing, but your setup might differ.

**Comment by matteoguarrera** [comment]
Is anyone willing to share their implementation to reproduce the results of the paper? :pleading_face: :pray:

**Comment by abdulfatir** [comment]
@matteoguarrera Unfortunately, despite trying for several weeks, I wasn't able to reproduce a number anywhere close to what's reported in the paper. Finally, for these datasets, I just copied the number from prior works in our paper: https://arxiv.org/abs/2301.11308

**Comment by nkiyohara** [comment]
Hello,

I’ve also attempted to reproduce the results of the CMU Mocap dataset as reported in "Scalable Gradients..." (Li et al., 2020). My best results hovered around an MSE of 10-20, which is still far from the reported Test MSE of 4.03 ± 0.20 in the paper.

To contribute to further discussions, I’ve uploaded my implementation here: [GitHub Repo Link](https://github.com/nkiyohara/latent-sde-mocap).

Feel free to review it and let me know if you spot any potential mistakes or areas for improvement in my implementation.
